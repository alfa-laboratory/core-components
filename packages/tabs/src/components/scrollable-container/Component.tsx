import React, { ReactNode, useEffect } from 'react';
import cn from 'classnames';
import computeScrollIntoView from 'compute-scroll-into-view';

import styles from './index.module.css';

export type ScrollableContainerProps = {
    /**
     * Дополнительный класс контейнера
     */
    containerClassName?: string;

    /**
     * Дочерние компоненты
     */
    children: ReactNode;

    /**
     * Активный элемент (всегда будет в видимой области)
     */
    activeChild: HTMLElement | null;
};

export const ScrollableContainer = ({
    containerClassName,
    children,
    activeChild,
}: ScrollableContainerProps) => {
    useEffect(() => {
        if (activeChild) {
            const actions = computeScrollIntoView(activeChild, {
                scrollMode: 'if-needed',
                block: 'nearest',
                inline: 'nearest',
            });

            // TODO: animate?
            actions.forEach(({ el, left }) => {
                // eslint-disable-next-line no-param-reassign
                el.scrollLeft = left;
            });
        }
    }, [activeChild]);

    return <div className={cn(styles.container, containerClassName)}>{children}</div>;
};
