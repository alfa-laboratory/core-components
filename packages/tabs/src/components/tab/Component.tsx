import React from 'react';
import cn from 'classnames';
import { TabProps } from '../../typings';

import styles from './index.module.css';

export const Tab = ({ children, hidden, className }: TabProps) => (
    <div
        className={cn(
            styles.component,
            {
                [styles.hidden]: hidden,
            },
            className,
        )}
        hidden={hidden}
        role='tabpanel'
        tabIndex={0}
    >
        {children}
    </div>
);
