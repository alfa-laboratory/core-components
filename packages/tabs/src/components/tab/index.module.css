.component {
    outline: none;
}

.hidden {
    display: none;
}
