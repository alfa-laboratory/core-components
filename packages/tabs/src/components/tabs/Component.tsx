import React, { cloneElement } from 'react';
import { TabsProps } from '../../typings';

export const Tabs = ({
    TabList,
    className,
    containerClassName,
    size,
    children,
    selectedId,
    scrollable,
    keepMounted = false,
    dataTestId,
    onChange,
}: Omit<TabsProps, 'view'>) => {
    const tabsArray = React.Children.toArray(children) as TabsProps['children'];
    const titles = tabsArray.map(({ props: { title, id, rightAddons } }) => ({
        title,
        id,
        rightAddons,
    }));
    const tabs = tabsArray.filter(
        tab => tab.props.id === selectedId || tab.props.keepMounted || keepMounted,
    );

    return (
        <div className={className}>
            <TabList
                containerClassName={containerClassName}
                size={size}
                titles={titles}
                selectedId={selectedId}
                scrollable={scrollable}
                onChange={onChange}
                dataTestId={dataTestId}
            />

            {tabs.map(tab => cloneElement(tab, { hidden: tab.props.id !== selectedId }))}
        </div>
    );
};
