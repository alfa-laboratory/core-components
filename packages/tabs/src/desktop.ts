export * from './components/primary-tablist/Component.desktop';
export * from './components/scrollable-container';
export * from './components/secondary-tablist/Component.desktop';
export * from './components/tab';
export * from './components/tabs/Component.desktop';
export * from './useTabs';
export * from './typings';
