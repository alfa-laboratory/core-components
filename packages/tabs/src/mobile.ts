export * from './components/primary-tablist/Component.mobile';
export * from './components/scrollable-container';
export * from './components/secondary-tablist/Component.mobile';
export * from './components/tab';
export * from './components/tabs/Component.mobile';
export * from './useTabs';
export * from './typings';
