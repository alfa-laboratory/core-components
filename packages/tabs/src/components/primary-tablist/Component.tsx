import React, { useEffect, useRef } from 'react';
import cn from 'classnames';
import { KeyboardFocusable } from '@alfalab/core-components-keyboard-focusable';
import { useTabs } from '../../useTabs';
import { ScrollableContainer } from '../scrollable-container';
import { TabListProps, Styles } from '../../typings';

export const PrimaryTabList = ({
    size,
    styles = {},
    className,
    containerClassName,
    titles = [],
    selectedId = titles.length ? titles[0].id : undefined,
    scrollable = true,
    onChange,
    dataTestId,
}: TabListProps & Styles) => {
    const lineRef = useRef<HTMLDivElement>(null);

    const { selectedTab, focusedTab, getTabListItemProps } = useTabs({
        titles,
        selectedId,
        onChange,
    });

    useEffect(() => {
        if (selectedTab && lineRef.current) {
            lineRef.current.style.width = `${selectedTab.offsetWidth}px`;
            lineRef.current.style.transform = `translateX(${selectedTab.offsetLeft}px)`;
        }
    });

    const renderContent = () => (
        <React.Fragment>
            {titles.map((item, index) => (
                <KeyboardFocusable key={item.id}>
                    {(ref, focused) => (
                        <button
                            {...getTabListItemProps(index, ref)}
                            type='button'
                            className={cn(styles.title, {
                                [styles.selected]: item.id === selectedId,
                            })}
                        >
                            <span className={cn(focused && styles.focused)}>{item.title}</span>
                            {item.rightAddons && (
                                <span className={styles.rightAddons}>{item.rightAddons}</span>
                            )}
                        </button>
                    )}
                </KeyboardFocusable>
            ))}

            <div className={styles.line} ref={lineRef} />
        </React.Fragment>
    );

    return (
        <div
            role='tablist'
            data-test-id={dataTestId}
            className={cn(styles.component, className, size && styles[size], {
                [styles.scrollable]: scrollable,
            })}
        >
            {scrollable ? (
                <ScrollableContainer
                    activeChild={focusedTab || selectedTab}
                    containerClassName={containerClassName}
                >
                    {renderContent()}
                </ScrollableContainer>
            ) : (
                <div className={cn(styles.container, containerClassName)}>{renderContent()}</div>
            )}
        </div>
    );
};
