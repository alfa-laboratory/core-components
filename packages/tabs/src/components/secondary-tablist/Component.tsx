import React from 'react';
import cn from 'classnames';
import { Tag } from '@alfalab/core-components-tag';
import { ScrollableContainer } from '../scrollable-container';
import { SecondaryTabListProps, Styles } from '../../typings';
import { useTabs } from '../../useTabs';

export const SecondaryTabList = ({
    styles = {},
    className,
    containerClassName,
    size,
    titles = [],
    selectedId = titles.length ? titles[0].id : undefined,
    scrollable = true,
    tagSize = 'xs',
    onChange,
    dataTestId,
}: SecondaryTabListProps & Styles) => {
    const { focusedTab, selectedTab, getTabListItemProps } = useTabs({
        titles,
        selectedId,
        onChange,
    });

    const renderContent = () =>
        titles
            .filter(item => !item.hidden)
            .map((item, index) => (
                <Tag
                    {...getTabListItemProps(index)}
                    key={item.id}
                    className={styles.title}
                    checked={item.id === selectedId}
                    size={tagSize}
                    rightAddons={item.rightAddons}
                >
                    {item.title}
                </Tag>
            ));

    return (
        <div
            role='tablist'
            data-test-id={dataTestId}
            className={cn(styles.component, className, size && styles[size])}
        >
            {scrollable ? (
                <ScrollableContainer
                    activeChild={focusedTab || selectedTab}
                    containerClassName={containerClassName}
                >
                    {renderContent()}
                </ScrollableContainer>
            ) : (
                <div className={cn(styles.container, containerClassName)}>{renderContent()}</div>
            )}
        </div>
    );
};
