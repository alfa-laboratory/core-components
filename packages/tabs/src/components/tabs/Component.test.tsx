/* eslint-disable multiline-comment-style */
import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import { TabsDesktop } from './Component.desktop';
import { TabsMobile } from './Component.mobile';
import { TabsResponsive } from './Component.responsive';
import { Tab } from '../tab';
import { TabsProps } from '../../typings';

const tabVariants: Array<[
    typeof TabsMobile | typeof TabsDesktop | typeof TabsResponsive,
    TabsProps['view'],
]> = [
    [TabsDesktop, 'primary'],
    [TabsDesktop, 'secondary'],
    [TabsMobile, 'primary'],
    [TabsMobile, 'secondary'],
    [TabsResponsive, 'primary'],
    [TabsResponsive, 'secondary'],
];

const renderTabs = (
    Component: typeof TabsMobile | typeof TabsDesktop | typeof TabsResponsive,
    props = {},
) =>
    render(
        <Component selectedId='tab-2' {...props}>
            <Tab title='Таб 1' id='tab-1' rightAddons='addon'>
                Таб 1
            </Tab>
            <Tab title='Таб 2' id='tab-2'>
                Таб 2
            </Tab>
            <Tab title='Таб 3' id='tab-3'>
                Таб 3
            </Tab>
            <Tab title='Таб 4' id='tab-4'>
                Таб 4
            </Tab>
            <Tab title='Таб 5' id='tab-5'>
                Таб 5
            </Tab>
        </Component>,
    );

describe('Tabs', () => {
    describe('Snapshots tests', () => {
        it.each(tabVariants)('should match snapshot', (Component, view) => {
            const { container } = renderTabs(Component, { view });
            expect(container).toMatchSnapshot();
        });
    });

    describe('Classes tests', () => {
        it.each(tabVariants)('should set custom class', (Component, view) => {
            const className = 'custom-class';
            const { container } = renderTabs(Component, { view, className });

            expect(container.firstElementChild).toHaveClass(className);
        });

        it.each(tabVariants)('should set custom container class', (Component, view) => {
            const containerClassName = 'custom-container-class';
            const { container } = renderTabs(Component, { view, containerClassName });

            expect(container.querySelector('.container')).toHaveClass(containerClassName);
        });
    });

    describe('Attributes tests', () => {
        it.each(tabVariants)('should set `data-test-id` atribute', (Component, view) => {
            const dataTestId = 'test-id';
            const { getByTestId } = renderTabs(Component, { view, dataTestId });

            expect(getByTestId(dataTestId).getAttribute('role')).toBe('tablist');
        });
    });

    describe('Render tests', () => {
        it.each(tabVariants)(
            'should render only selected tabpanel by default',
            (Component, view) => {
                const { getAllByRole } = renderTabs(Component, { view });
                const tabpanels = getAllByRole('tabpanel');

                expect(tabpanels.length).toEqual(1);
            },
        );

        it.each(tabVariants)(
            'should render only one visible tabpanel if keepMounted=`true`',
            (Component, view) => {
                const { getAllByRole } = renderTabs(Component, { view, keepMounted: true });
                const hidden = getAllByRole('tabpanel', { hidden: true });
                const visible = getAllByRole('tabpanel', { hidden: false });

                expect(hidden.length).toBeGreaterThan(1);
                expect(visible.length).toEqual(1);
            },
        );

        it.each(tabVariants)('should render corresponding tabpanel', (Component, view) => {
            const { getByRole } = renderTabs(Component, { view });
            const selectedTab = getByRole('tab', { selected: true });
            const tabpanel = getByRole('tabpanel', { hidden: false });

            expect(tabpanel.textContent).toEqual(selectedTab.textContent);
        });

        it.each(tabVariants)('should unmount without errors', (Component, view) => {
            const { unmount } = renderTabs(Component, { view });

            expect(unmount).not.toThrowError();
        });

        it('should not render empty tabs', () => {
            const { queryAllByRole } = render(
                <TabsDesktop selectedId='tab-2' keepMounted={true}>
                    <Tab title='Таб 1' id='tab-1' />
                    <Tab title='Таб 2' id='tab-2' />
                    <Tab title='Таб 3' id='tab-3'>
                        Таб 3
                    </Tab>
                </TabsDesktop>,
            );

            expect(queryAllByRole('tabpanel', { hidden: true })).toHaveLength(1);
            expect(queryAllByRole('tabpanel', { hidden: false })).toHaveLength(0);
        });
    });

    describe('Interaction tests', () => {
        it.each(tabVariants)('should call `onChange`', (Component, view) => {
            const cb = jest.fn();
            const { getAllByRole } = renderTabs(Component, { view, onChange: cb });
            const tab = getAllByRole('tab', { selected: false });

            if (tab.length > 0) {
                fireEvent.click(tab[0]);
            }

            expect(cb).toBeCalledTimes(1);
        });

        it.each(tabVariants)('should not call `onChange` for selected tab', (Component, view) => {
            const cb = jest.fn();
            const { getByRole } = renderTabs(Component, { view, onChange: cb });
            const tab = getByRole('tab', { selected: true });

            if (tab) {
                fireEvent.click(tab);
            }

            expect(cb).not.toBeCalled();
        });
    });
});
