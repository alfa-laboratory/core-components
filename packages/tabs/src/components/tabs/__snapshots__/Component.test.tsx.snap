// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs Snapshots tests should match snapshot 1`] = `
<div>
  <div>
    <div
      class="component m"
      role="tablist"
    >
      <div
        class="container"
      >
        <button
          aria-selected="false"
          class="title"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span
            class=""
          >
            Таб 1
          </span>
          <span
            class="rightAddons"
          >
            addon
          </span>
        </button>
        <button
          aria-selected="true"
          class="title selected"
          role="tab"
          tabindex="0"
          type="button"
        >
          <span
            class=""
          >
            Таб 2
          </span>
        </button>
        <button
          aria-selected="false"
          class="title"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span
            class=""
          >
            Таб 3
          </span>
        </button>
        <button
          aria-selected="false"
          class="title"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span
            class=""
          >
            Таб 4
          </span>
        </button>
        <button
          aria-selected="false"
          class="title"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span
            class=""
          >
            Таб 5
          </span>
        </button>
        <div
          class="line"
          style="width: 0px; transform: translateX(0px);"
        />
      </div>
    </div>
    <div
      class="component"
      role="tabpanel"
      tabindex="0"
    >
      Таб 2
    </div>
  </div>
</div>
`;

exports[`Tabs Snapshots tests should match snapshot 2`] = `
<div>
  <div>
    <div
      class="component m"
      role="tablist"
    >
      <div
        class="container"
      >
        <button
          aria-selected="false"
          class="component s title"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span>
            Таб 1
          </span>
          <span
            class="addons"
          >
            addon
          </span>
        </button>
        <button
          aria-selected="true"
          class="component s checked title"
          role="tab"
          tabindex="0"
          type="button"
        >
          <span>
            Таб 2
          </span>
        </button>
        <button
          aria-selected="false"
          class="component s title"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span>
            Таб 3
          </span>
        </button>
        <button
          aria-selected="false"
          class="component s title"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span>
            Таб 4
          </span>
        </button>
        <button
          aria-selected="false"
          class="component s title"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span>
            Таб 5
          </span>
        </button>
      </div>
    </div>
    <div
      class="component"
      role="tabpanel"
      tabindex="0"
    >
      Таб 2
    </div>
  </div>
</div>
`;

exports[`Tabs Snapshots tests should match snapshot 3`] = `
<div>
  <div>
    <div
      class="undefined"
      role="tablist"
    >
      <div
        class="component"
      >
        <div
          class="container"
        >
          <button
            aria-selected="false"
            class=""
            role="tab"
            tabindex="-1"
            type="button"
          >
            <span
              class=""
            >
              Таб 1
            </span>
            <span>
              addon
            </span>
          </button>
          <button
            aria-selected="true"
            class="undefined"
            role="tab"
            tabindex="0"
            type="button"
          >
            <span
              class=""
            >
              Таб 2
            </span>
          </button>
          <button
            aria-selected="false"
            class=""
            role="tab"
            tabindex="-1"
            type="button"
          >
            <span
              class=""
            >
              Таб 3
            </span>
          </button>
          <button
            aria-selected="false"
            class=""
            role="tab"
            tabindex="-1"
            type="button"
          >
            <span
              class=""
            >
              Таб 4
            </span>
          </button>
          <button
            aria-selected="false"
            class=""
            role="tab"
            tabindex="-1"
            type="button"
          >
            <span
              class=""
            >
              Таб 5
            </span>
          </button>
          <div
            style="width: 0px; transform: translateX(0px);"
          />
        </div>
      </div>
    </div>
    <div
      class="component"
      role="tabpanel"
      tabindex="0"
    >
      Таб 2
    </div>
  </div>
</div>
`;

exports[`Tabs Snapshots tests should match snapshot 4`] = `
<div>
  <div>
    <div
      class="undefined"
      role="tablist"
    >
      <div
        class="component"
      >
        <div
          class="container"
        >
          <button
            aria-selected="false"
            class="component xs"
            role="tab"
            tabindex="-1"
            type="button"
          >
            <span>
              Таб 1
            </span>
            <span
              class="addons"
            >
              addon
            </span>
          </button>
          <button
            aria-selected="true"
            class="component xs checked"
            role="tab"
            tabindex="0"
            type="button"
          >
            <span>
              Таб 2
            </span>
          </button>
          <button
            aria-selected="false"
            class="component xs"
            role="tab"
            tabindex="-1"
            type="button"
          >
            <span>
              Таб 3
            </span>
          </button>
          <button
            aria-selected="false"
            class="component xs"
            role="tab"
            tabindex="-1"
            type="button"
          >
            <span>
              Таб 4
            </span>
          </button>
          <button
            aria-selected="false"
            class="component xs"
            role="tab"
            tabindex="-1"
            type="button"
          >
            <span>
              Таб 5
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      class="component"
      role="tabpanel"
      tabindex="0"
    >
      Таб 2
    </div>
  </div>
</div>
`;

exports[`Tabs Snapshots tests should match snapshot 5`] = `
<div>
  <div>
    <div
      class="component m"
      role="tablist"
    >
      <div
        class="container"
      >
        <button
          aria-selected="false"
          class="title"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span
            class=""
          >
            Таб 1
          </span>
          <span
            class="rightAddons"
          >
            addon
          </span>
        </button>
        <button
          aria-selected="true"
          class="title selected"
          role="tab"
          tabindex="0"
          type="button"
        >
          <span
            class=""
          >
            Таб 2
          </span>
        </button>
        <button
          aria-selected="false"
          class="title"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span
            class=""
          >
            Таб 3
          </span>
        </button>
        <button
          aria-selected="false"
          class="title"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span
            class=""
          >
            Таб 4
          </span>
        </button>
        <button
          aria-selected="false"
          class="title"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span
            class=""
          >
            Таб 5
          </span>
        </button>
        <div
          class="line"
          style="width: 0px; transform: translateX(0px);"
        />
      </div>
    </div>
    <div
      class="component"
      role="tabpanel"
      tabindex="0"
    >
      Таб 2
    </div>
  </div>
</div>
`;

exports[`Tabs Snapshots tests should match snapshot 6`] = `
<div>
  <div>
    <div
      class="component m"
      role="tablist"
    >
      <div
        class="container"
      >
        <button
          aria-selected="false"
          class="component s title"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span>
            Таб 1
          </span>
          <span
            class="addons"
          >
            addon
          </span>
        </button>
        <button
          aria-selected="true"
          class="component s checked title"
          role="tab"
          tabindex="0"
          type="button"
        >
          <span>
            Таб 2
          </span>
        </button>
        <button
          aria-selected="false"
          class="component s title"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span>
            Таб 3
          </span>
        </button>
        <button
          aria-selected="false"
          class="component s title"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span>
            Таб 4
          </span>
        </button>
        <button
          aria-selected="false"
          class="component s title"
          role="tab"
          tabindex="-1"
          type="button"
        >
          <span>
            Таб 5
          </span>
        </button>
      </div>
    </div>
    <div
      class="component"
      role="tabpanel"
      tabindex="0"
    >
      Таб 2
    </div>
  </div>
</div>
`;
